book/src/feature/stream.md

```diff
@@ -19,3 +19,36 @@ See [logging examples](logging.md) which uses streams extensively.
 ## What about streaming from Dart/Flutter to Rust?
 
 This is not currently supported. As a workaround, consider iterating through your Dart stream and calling a normal Rust function for each item.
+
+## Streaming external types (mirror)
+
+When using mirrored external types in a StreamSink you have to specify the generated `mirror_..` type in the StreamSink. This is because the generated code can not implement the necessary interfaces on a struct from a different crate. The generated `mirror_struct` implements `From<struct>` so you can use your normal types and just call `.into()` before passing them to the StreamSink.
```

Comment:
Ok I start reviewing this PR. Good job! Though I am a bit hesitate about this developer interface. Is it possible we do not leak the "mirror_something" types to end users? That looks like internal implementation details, and we should hide them.

Reply:
The problem is that Stream sink requires the inner type to implement IntoDart and we can not implement that on external types.
Maybe we could require the type for the stream to implement Into<> for some type that implements IntoDart. But I am not sure that is possible.
We could also hide the mirror type with a macro that converts from struct to mirror_struct but I don't think that is much better.

Something like StreamSink<mirror!(ExternalStruct)>

Reply:
Yes I agree that's not trivial to solve.

For brainstom, some more naive thoughts (may not be good): We may implement StreamSink to convert the types internally (from ExternalStruct to mirror_ExternalStruct).

But we do not want to change signature visible to the user (e.g. `StreamSink<ExternalStruct, mirror_ExternalStruct>` does not sound very beautiful.

Then maybe

```rust
// static code
trait StreamSink<T> { ... }

// generated code
// since it is generated, we are free to generate the code that converts ExternalStruct into mirror_ExternalStruct, just like what we did for return values
struct _StreamSinkExternalStruct;
impl StreamSink<ExternalStruct> for _StreamSinkExternalStruct {...}

// again generated code
fn generated_bridge_function_that_calls_f() {
  f(_StreamSinkExternalStruct(...))
}

// user in api.rs
fn f(stream: impl StreamSink<ExternalStruct>) {}
```

Reply:
Also cc @Desdaemon in case you want to play with these :)

Reply:
I got something like that working (not generated yet, just typed out)

generated code:
```rust
pub trait LogEntryStreamSink {
    fn add(&self, val: LogEntry);
}
impl LogEntryStreamSink for StreamSink<LogEntry> {
    fn add(&self, val: LogEntry) {
        self.add_inner::<_, LogEntry>(val);
    }
}
```

api
```rust
use crate::bridge_generated::VecBleDeviceStreamSink;
pub fn create_log_stream(s: StreamSink<LogEntry>) {
    s.add(LogEntry {})
}

```

You still have to import a trait to get the method but we don't have the mirror_ type in the api anymore.
`add_inner` was previously the add method on a sink. I had to change some type constraints for this to work.

If this interface looks good i will try to generate the code automatically. I had to add some type annotations but only in the generated.rs file.

Reply:
That way of doing it would also make all other changes from this PR obsolete :D
I tried it on a new branch without any of the other changes I made so far.

Reply:
You proposal looks also interesting!

A bit of worry: What if the user wants to use StreamSink generically?
But that should not be a big problem, because the user can just call `add_inner` (and manually pass the type argument).

> I tried it on a new branch without any of the other changes I made so far.

Btw you can make a "draft PR" for that branch, b/c it will be easy to see `Files changed` with a (draft) PR